### PR TITLE
Search box for widget palette

### PR DESCRIPTION
**Pull request by aayush420**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -21,20 +22,40 @@ var (
...
+		s = strings.ToLower(s)
+		tempNames = []string{}
+		for i := 0; i < len(widgetLowerNames); i++ {
+			if strings.Contains(widgetLowerNames[i], s) {
+				tempNames = append(tempNames, widgetNames[i])
+			}
+		}
+		list.Refresh()
+	}
+	searchBox.SetText(" ")
```
Comment (andydotxyz) [inline main.go]: This looks like a bit of a workaround - why not set it up correctly to start:

```diff
diff --git a/main.go b/main.go
index 42bc4c3..e42d59d 100644
--- a/main.go
+++ b/main.go
@@ -26,6 +26,7 @@ func buildLibrary() fyne.CanvasObject {
        widgetLowerNames := []string{}
        for _, name := range widgetNames {
                widgetLowerNames = append(widgetLowerNames, strings.ToLower(name))
+               tempNames = append(tempNames, name)
        }
        list := widget.NewList(func() int {
                return len(tempNames)
@@ -52,8 +53,6 @@ func buildLibrary() fyne.CanvasObject {
                }
                list.Refresh()
        }
-       searchBox.SetText(" ")
-       searchBox.SetText("")
 
        return container.NewBorder(searchBox, widget.NewButtonWithIcon("Insert", theme.ContentAddIcon(), func() {
                if c, ok := current.(*overlayContainer); ok {
```

**Review by andydotxyz — CHANGES_REQUESTED** [review]
Excellent, though some of the code is a little strange - suggested improvement inline

**Review by andydotxyz — CHANGES_REQUESTED** [review]
This code looks great and runs well. 
However:
* If you select an item in the pallete
* search for something else
* new items get selected
* add item to a container
* - notice that different item (the old one) is added.

This may be caused by a bug in list, I'm not sure. If that is the case then we should de-select the list item before updating it's display...

**Comment by andydotxyz** [comment]
Please don't force-push after a review has been added. It means we lose all the context and have to start again.

**Comment by aayush420** [comment]
> Please don't force-push after a review has been added. It means we lose all the context and have to start again.

I was just trying to rebase the branch on the latest HEAD from main.

**Comment by andydotxyz** [comment]
> I was just trying to rebase the branch on the latest HEAD from main.

You should not rebase public branches (especially when attached to a PR) - it rewrites history. Use merge instead.

**Review by andydotxyz — CHANGES_REQUESTED** [review]
Almost! It now does not have the strange issue, but when nothing is selected it remembers the previous selection - which because of this code is sometimes not what you would think.
I think this should be resolved by adding:

```go
	list.OnUnselected = func(widget.ListItemID) {
		selected = nil
	}
```

**Review by andydotxyz — APPROVED** [review]
Great new feature, thanks

**Pull request merged**